api/app/Repositories/ArticleRepository.php

```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: redjik
+ * Date: 08.07.15
+ * Time: 23:40
+ */
+
+namespace App\Repositories;
+
+use App\Models\Article;
+use App\Models\ArticlePermalink;
+use Rhumsaa\Uuid\Uuid;
+
+class ArticleRepository extends BaseRepository
+{
+    /**
+     * @param string $id
+     * @param array $columns
+     * @return \App\Models\BaseModel|null
+     */
+    public function find($id, $columns = ['*'])
+    {
+        $builder = $this->model->query();
+        $query = $builder->getQuery();
+        $tableName = $this->model->getTable();
+        $joinTableName = ArticlePermalink::getTableName();
+        $query->join($joinTableName, $joinTableName.'.article_id', '=', $tableName.'.article_id', 'left');
+        if (Uuid::isValid($id)) {
+            $query->where($tableName.'.article_id', '=', $id);
+        }
+
+        $query->orWhere($joinTableName.'.uri', '=', $id);
+
+        return $builder->get($columns)->first();
```

Comment:
This function needs to be optimised by first attempting to check by Article->articleId or Article->permalink. Only if that fails should the permalink history be checked. Article's being accessed by their permalink history is a rare case so the join should not be completed for every access for performance reasons.

e.g.
```
try {
   return Article::find($id)->orWhere('permalink', '=', $id)->firstOrFail();
}catch(ModelNotFoundException){ //id or permalink not found, try permalink history
   return ArticlePermalink::find($id)->article();
}
```
I haven't tested the above code so you may need to modify it somewhat. Also in general use the eloquent methods over using the query builder. In general my practice is to try to avoid using the query builder. Sometimes if a query is getting too complex it is actually a good indicator that there needs to be better relations set up, or a reporting sql view created.